### PR TITLE
refactor: various cleanups and fixes in `chain.rs`

### PR DESCRIPTION
**Pull request by akhi3030**
get_state_response_part():
- remove the contiguous shard_ids assumption.  Instead of checking the shard id against the number of chunks, look up the list of shard ids instead.  
- since the epoch id is remains the same for the current and the prev block, the set of shard ids should also remain the same so we only need to do the check once.
- remove some unnecessary clones
- finally rename some variables to make them shorter.  Mainly remove the `sync_` prefix.  I am not sure how much value this prefix is adding and the shorter variables are still conveying the same amount of information.

compute_state_response_header():
- Similar to above, remove the contiguous shard_ids assumption.  
- Remove some unnecessary clones and dereferences

save_receipt_id_to_shard_id_for_block():
- there is a potential bug in the function where if encounter an error in middle of iteration, we do not roll back the changes.  Hence, pre compute the list of hashmaps before saving receipts.
- Take a list of shard ids instead of num_shards to remove the contiguous shard ids assumption

### CODE REVIEW

**Comment by nikurt** [comment]
Thank you for the cleanup, the change looks good.

I scanned the file.
I found a similar pattern in the following related lines:
* https://github.com/near/nearcore/blob/8cd5138f8ac920e10eecb12127edc17b71cba1ec/chain/chain/src/chain.rs#L2916
* https://github.com/near/nearcore/blob/8cd5138f8ac920e10eecb12127edc17b71cba1ec/chain/chain/src/chain.rs#L2926
* https://github.com/near/nearcore/blob/8cd5138f8ac920e10eecb12127edc17b71cba1ec/chain/chain/src/chain.rs#L2953

And one more unrelated assumption of consecutive shard ids:
* https://github.com/near/nearcore/blob/8cd5138f8ac920e10eecb12127edc17b71cba1ec/chain/chain/src/chain.rs#L4985

Would you mind cleaning them up too?

**Comment by akhi3030** [comment]
fixed the additional issues as well.

**Comment by akhi3030** [comment]
@nikurt: doesn't look like I can re-request a review from you yet.  Please treat this as a ping to review.

File: chain/chain/src/chain.rs [hunk 1]
```diff
@@ -2949,9 +2947,6 @@ impl Chain {
             .get_block(block_header.prev_hash())
         {
             Ok(prev_block) => {
-                if shard_id as usize >= prev_block.chunks().len() {
```
Comment (akhi3030) [inline chain/chain/src/chain.rs:2952]: Note to reviewer: I am not 100% convinced that this is correct so please double check that this is fine.

Comment (nikurt) [reply chain/chain/src/chain.rs:2952]: Seems correct to me.
I'm more worried about using the same `shard_id` value for two different shard layouts: prev epoch and the `sync_hash` epoch.
Fixing that is out of scope of your PR.

**Pull request merged**